src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -137,11 +138,41 @@ public IList<ArraySegment<byte>> BufferList
                 StartConfiguring();
                 try
                 {
-                    if (value != null && _buffer != null)
+                    if (value != null)
                     {
-                        throw new ArgumentException(SR.Format(SR.net_ambiguousbuffers, "Buffer"));
+                        if (_buffer != null)
+                        {
+                            // Can't have both set
+                            throw new ArgumentException(SR.Format(SR.net_ambiguousbuffers, "Buffer"));
+                        }
+
+                        // Copy the user-provided list into our internal buffer list,
+                        // so that we are not affected by subsequent changes to the list.
+                        // We reuse the existing list so that we can avoid reallocation when possible.
+                        if (_bufferListInternal == null)
+                        {
+                            _bufferListInternal = new List<ArraySegment<byte>>(value.Count);
+                        }
+                        else
+                        {
+                            _bufferListInternal.Clear();
+                        }
+
+                        for (int i = 0; i < value.Count; i++)
+                        {
+                            _bufferListInternal.Add(value[i]);
+                        }
```

Comment:
Alternatively you could just store an array and a count rather than allocating a list, using a for loop and Add and the interface indexer instead of CopyTo, etc.